src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -513,14 +589,23 @@ void recordWriteAfterQueryExecution() throws SQLException {
         }
     }
 
-    public Statement getReadStatement(String sql) {
+    public Statement getReadStatement(String sql, QueryContext.Builder contextBuilder) {
         isWriteOperation = isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
+        if (isWriteOperation) {
+            contextBuilder.reason(WRITE_OPERATION).runOnMain(true);
+            return prepareWriteStatement();
+        }
+        if (connectionProvider.getState().equals(MAIN)) {
+            contextBuilder.reason(MAIN_CONNECTION_REUSE).runOnMain(true);
+            return prepareWriteStatement();
+        }
 
-        // if write connection is already initialized, but the current statement is null
-        // we should use a write statement, regardless of the fact that readStatement has been called.
-        if (connectionProvider.getState().equals(MAIN) || isWriteOperation || isSelectForUpdate(sql)) {
+        if (isSelectForUpdate(sql)) {
+            contextBuilder.reason(LOCK).runOnMain(true);
             return prepareWriteStatement();
         }
+
+        readStatement.setContextBuilder(contextBuilder);
```

Comment:
This is where it would use the distant `READ_OPERATION` de facto default

Reply:
The default decision could bring `sql`, so it cannot be inlined here. Gotta let the `getReadStatement(RouteDecisionBuilder decisionBuilder)` parameter be